src/set/set.service.ts

```diff
@@ -1,345 +1,418 @@
+        taskId: ObjectId,
+        deleteType: string,
+        user: JwtUserDto,
+    ): Promise<UpdatedCounts> {
+        // Hard delete
+        if (deleteType === 'hard') {
+            if (user.role != 'admin') throw new ForbiddenException()
+
+            const set = await this.setSchema.findOneAndUpdate(
+                { _id: setId, 'tasks._id': taskId },
+                { $pull: { tasks: { _id: taskId } } },
+            )
+
+            if (!set) {
+                throw new NotFoundException()
+            }
 
-    return updatedResult
+            const updatedResult = await this.updateCounts(setId)
 
-  }
+            return updatedResult
+        }
 
+        // Soft delete
+        const queryMatch: {
+            _id: ObjectId
+            'tasks._id': ObjectId
+            createdBy?: ObjectId
+        } = { _id: setId, 'tasks._id': taskId }
+        if (user.role !== Role.Admin) queryMatch.createdBy = user.userId
 
-  /*------------------------------------\
-  |               Helpers               |
-  \------------------------------------*/
+        const set = await this.setSchema.findOneAndUpdate(queryMatch, {
+            'tasks.$.status': Status.DELETED,
+        })
 
+        if (!set) throw new NotFoundException()
 
-  private onlyActiveTasks(set: (ResponseSet & { tasks: ResponseTaskWithStatus[] })): ResponseSetWithTasks {
+        const updatedResult = await this.updateCounts(setId)
 
-    const reducedTasks: ResponseTask[] = [];
+        return updatedResult
+    }
 
-    // Iterate over the tasks array and only push those that are active
-    set.tasks.forEach((task) => {
-      if (task.status === Status.ACTIVE) {
-        reducedTasks.push({
-          currentPlayerGender: task.currentPlayerGender,
-          _id: task._id,
-          type: task.type,
-          message: task.message
+    // Helpers
+
+    private onlyActiveTasks(
+        set: ResponseSet & { tasks: ResponseTaskWithStatus[] },
+    ): ResponseSetWithTasks {
+        const reducedTasks: ResponseTask[] = []
+
+        // Iterate over the tasks array and only push those that are active
+        set.tasks.forEach((task) => {
+            if (task.status === Status.ACTIVE) {
+                reducedTasks.push({
+                    currentPlayerGender: task.currentPlayerGender,
+                    _id: task._id,
+                    type: task.type,
+                    message: task.message,
+                })
+            }
         })
-      }
-    })
 
-    return {
-      ...set,
-      tasks: reducedTasks
+        return {
+            ...set,
+            tasks: reducedTasks,
+        }
     }
-  }
 
-  // Uses 2 additional database calls to update the task counts and return the new settings
-  private async updateCounts(setId: ObjectId): Promise<UpdatedCounts> {
+    // Uses 2 additional database calls to update the task counts and return the new settings
+    private async updateCounts(setId: ObjectId): Promise<UpdatedCounts> {
```

Comment:
This probably is the best way to do this but this also is confusing. So either comment or create a short wiki entry to make understanding this easier :)